File: src/api/forum.ts
import { api } from "./instance";

export interface ForumAuthor {
  id: number;
  fullName: string;
  avatar: string;
  stars: number;
  role: "admin" | "teacher" | "student" | "parent";
}

export interface ForumCategory {
  id: number;
  name: string;
  description: string;
  topicsCount: number;
}

export interface ForumTopic {
  id: number;
  title: string;
  content: string;
  author: ForumAuthor;
  categoryId: number;
  views: number;
  isSolved: boolean;
  createdAt: string;
  replies: number;
}

export interface ForumComment {
  id: number;
  content: string;
  author: ForumAuthor;
  isSolution: boolean;
  createdAt: string;
}

interface TopicDetail {
  topic: ForumTopic;
  comments: ForumComment[] | null;
}

interface CreateTopicBody {
  title: string;
  content: string;
  categoryId: number;
}

interface CreateCategoryBody {
  name: string;
  description: string;
  order: number;
}

export const getForumCategories = () => api<ForumCategory[] | null>("/forum/categories");

export const createForumCategory = (body: CreateCategoryBody) =>
  api<ForumCategory>("/forum/categories", { method: "POST", body });

export const getForumTopics = (categoryId?: number) =>
  api<ForumTopic[] | null>("/forum/topics", {
    query: categoryId ? { categoryId } : undefined,
  });

export const getForumTopic = (id: number) => api<TopicDetail>(`/forum/topics/${id}`);

export const createForumTopic = (body: CreateTopicBody) =>
  api<ForumTopic>("/forum/topics", { method: "POST", body });

export const createForumComment = (topicId: number, content: string) =>
  api<ForumComment>(`/forum/topics/${topicId}/comments`, {
    method: "POST",
    body: { content },
  });

export const markSolution = (commentId: number, topicId: number) =>
  api(`/forum/comments/${commentId}/solution`, {
    method: "PUT",
    query: { topicId },
  });

File: api/internal/api/presenter/forum.go
package presenter

import (
	"time"
)

type ForumAuthor struct {
	ID       uint   `json:"id"`
	FullName string `json:"fullName"`
	Avatar   string `json:"avatar"`
	Stars    int    `json:"stars"`
	Role     string `json:"role"`
}

type ForumCategoryResponse struct {
	ID          uint   `json:"id"`
	Name        string `json:"name"`
	Description string `json:"description"`
	TopicsCount int    `json:"topicsCount"`
}

type ForumTopicResponse struct {
	ID         uint        `json:"id"`
	Title      string      `json:"title"`
	Content    string      `json:"content"`
	Author     ForumAuthor `json:"author"`
	CategoryID uint        `json:"categoryId"`
	Views      int         `json:"views"`
	IsSolved   bool        `json:"isSolved"`
	CreatedAt  time.Time   `json:"createdAt"`
	Replies    int         `json:"replies"`
}

type ForumTopicDetailResponse struct {
	Topic    ForumTopicResponse     `json:"topic"`
	Comments []ForumCommentResponse `json:"comments"`
}

type ForumCommentResponse struct {
	ID         uint        `json:"id"`
	Content    string      `json:"content"`
	Author     ForumAuthor `json:"author"`
	IsSolution bool        `json:"isSolution"`
	CreatedAt  time.Time   `json:"createdAt"`
}

type CreateForumTopicRequest struct {
	Title      string `json:"title" example:"Как использовать GORM?"`
	Content    string `json:"content" example:"У меня вопрос по связям..."`
	CategoryID uint   `json:"categoryId" example:"1"`
}

type CreateForumCategoryRequest struct {
	Name        string `json:"name" example:"Вопросы по Go"`
	Description string `json:"description" example:"Обсуждение языка программирования Go"`
	Order       int    `json:"order" example:"1"`
}

type CreateForumCommentRequest struct {
	Content string `json:"content" example:"Вот решение вашей проблемы..."`
}
